rules/terraformrules/terraform_empty_list_equality.go

```diff
@@ -52,30 +52,43 @@ func (r *TerraformEmptyListEqualityRule) Check(runner *tflint.Runner) error {
 	return nil
 }
 
+// checkEmptyList visits all blocks that can contain expressions and checks for comparisons with static empty list
 func (r *TerraformEmptyListEqualityRule) checkEmptyList(runner *tflint.Runner) error {
 	return runner.WalkExpressions(func(expr hcl.Expression) error {
 		if conditionalExpr, ok := expr.(*hclsyntax.ConditionalExpr); ok {
-			if binaryOpExpr, ok := conditionalExpr.Condition.(*hclsyntax.BinaryOpExpr); ok {
-				if binaryOpExpr.Op.Type.FriendlyName() == "bool" {
-					if right, ok := binaryOpExpr.RHS.(*hclsyntax.TupleConsExpr); ok {
-						checkEmptyList(right, runner, r, binaryOpExpr)
-					}
-					if left, ok := binaryOpExpr.LHS.(*hclsyntax.TupleConsExpr); ok {
-						checkEmptyList(left, runner, r, binaryOpExpr)
-					}
-				}
-			}
+			var issuesRangeSet = make(map[hcl.Range]struct{})
+			r.searchEmptyList(conditionalExpr.Condition, runner, conditionalExpr.Range(), issuesRangeSet)
+			r.emitEmptyListEqualityIssues(issuesRangeSet, runner)
 		}
 		return nil
 	})
 }
 
-func checkEmptyList(tupleConsExpr *hclsyntax.TupleConsExpr, runner *tflint.Runner, r *TerraformEmptyListEqualityRule, binaryOpExpr *hclsyntax.BinaryOpExpr) {
-	if len(tupleConsExpr.Exprs) == 0 {
+// searchEmptyList Searches for comparisons with static empty list in the given expression
+func (r *TerraformEmptyListEqualityRule) searchEmptyList(expr hcl.Expression, runner *tflint.Runner, exprRange hcl.Range, issuesRangeSet map[hcl.Range]struct{}) {
+	if binaryOpExpr, ok := expr.(*hclsyntax.BinaryOpExpr); ok {
+		if binaryOpExpr.Op.Type.FriendlyName() == "bool" {
+			r.searchEmptyList(binaryOpExpr.RHS, runner, binaryOpExpr.Range(), issuesRangeSet)
+			r.searchEmptyList(binaryOpExpr.LHS, runner, binaryOpExpr.Range(), issuesRangeSet)
+		}
+	} else if binaryOpExpr, ok := expr.(*hclsyntax.BinaryOpExpr); ok {
+		r.searchEmptyList(binaryOpExpr, runner, binaryOpExpr.Range(), issuesRangeSet)
+	} else if parenthesesExpr, ok := expr.(*hclsyntax.ParenthesesExpr); ok {
+		r.searchEmptyList(parenthesesExpr.Expression, runner, parenthesesExpr.Range(), issuesRangeSet)
+	} else if tupleConsExpr, ok := expr.(*hclsyntax.TupleConsExpr); ok {
+		if len(tupleConsExpr.Exprs) == 0 {
+			issuesRangeSet[exprRange] = struct{}{}
+		}
+	}
+}
```

Comment:
Perhaps this search logic can be rewritten fairly easily by #1432.

Reply:
Hi @wata727 ! Thanks for the suggestion and sorry for being silent in the last couple of weeks. Would you mind to ellaborate on your suggestion please? 😕 I'm calling `WalkExpressions` [here](https://github.com/terraform-linters/tflint/pull/1426/files/dd24efe8c505b32e3d15b093295eddeb85d55605#diff-c4c162621b4beba8a29758aed0555eb9d6a2e326ae4b2f537f3fad72592432c4R57) but as you noticed the latest change didn't fix the issue, therefore I think I still need to process the expression recursively here.

Reply:
I think the pattern that this rule essentially wants to detect is `any == []`. In other words, it might be better to check against `BinaryOpExpr` while walking through all the expressions, rather than checking for `ConditionalExpr` first.

Imagine something like:

```go
runner.WalkExpression(func(expr hcl.Expression) error {
  if binaryOpExpr, ok := expr.(*hclsyntax.BinaryOpExpr); ok && binaryOpExpr.Op.Type == cty.Bool {
    if tupleConsExpr, ok := binaryOpExpr.RHS.(*hclsyntax.TupleConsExpr); ok && len(tupleConsExpr.Exprs) == 0 {
      runner.EmitIssue(...)
    }

    if tupleConsExpr, ok := binaryOpExpr.LHS.(*hclsyntax.TupleConsExpr); ok && len(tupleConsExpr.Exprs) == 0 {
      runner.EmitIssue(...)
    }
  }
})
```

What do you think?

Reply:
ahh! got it 🙇 Just tried it and it is much better now! I could even remove the map to avoid duplicated ranges 👌 thank you!